### PR TITLE
Add more client cert related bindings

### PR DESCRIPTION
**Pull request by reaperhulk**
replaces #1923

Pushing this up because jenkins failures on the old build have been deleted (and the libre check was incorrect).

### CODE REVIEW

File: src/cryptography/hazmat/bindings/openssl/engine.py [hunk 1]
```diff
@@ -145,11 +147,23 @@
...
 EVP_PKEY *ENGINE_load_private_key(ENGINE *, const char *, UI_METHOD *, void *);
 EVP_PKEY *ENGINE_load_public_key(ENGINE *, const char *, UI_METHOD *, void *);
+
 void ENGINE_add_conf_module(void);
 """
 
 MACROS = """
 void ENGINE_load_cryptodev(void);
+
+/* Engine's client cert functions were added in OpenSSL 0.9.8i */
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/engine.py]: Are you sure, I grepped and it looks like it's 0.9.8k

File: src/cryptography/hazmat/bindings/openssl/engine.py [hunk 2]
```diff
@@ -145,11 +147,23 @@
...
+int ENGINE_set_load_ssl_client_cert_function(ENGINE *,
+                                             ENGINE_SSL_CLIENT_CERT_PTR);
+ENGINE_SSL_CLIENT_CERT_PTR ENGINE_get_ssl_client_cert_function(
+    const ENGINE *);
+int ENGINE_load_ssl_client_cert(ENGINE *, SSL *,
+                                Cryptography_STACK_OF_X509_NAME *,
+                                X509 **, EVP_PKEY **,
+                                Cryptography_STACK_OF_X509 **,
+                                UI_METHOD *, void *);
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/engine.py]: This isn't right, as far as I can tell the signature for this function changed at some point over the years.

Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/engine.py]: This is teh original sig:

```diff
+int ENGINE_load_ssl_client_cert(ENGINE *e, SSL *s,
+       STACK_OF(X509_NAME) *ca_dn, X509 **pcert, EVP_PKEY **ppkey,
+       UI_METHOD *ui_method, void *callback_data)
```

File: src/cryptography/hazmat/bindings/openssl/ssl.py [hunk 3]
```diff
@@ -186,6 +188,14 @@
 X509 *SSL_get_peer_certificate(const SSL *);
 int SSL_get_ex_data_X509_STORE_CTX_idx(void);
 
+int SSL_use_certificate(SSL *, X509 *);
+int SSL_use_certificate_ASN1(SSL *, const unsigned char *, int);
+int SSL_use_certificate_file(SSL *, const char *, int);
+int SSL_use_PrivateKey(SSL *, EVP_PKEY *);
+int SSL_use_PrivateKey_ASN1(int,SSL *, const unsigned char *, long);
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/ssl.py]: Missing space here.

File: src/cryptography/hazmat/bindings/openssl/ssl.py [hunk 4]
```diff
@@ -387,6 +407,13 @@
 void SSL_get0_alpn_selected(const SSL *, const unsigned char **, unsigned *);
 
 long SSL_get_server_tmp_key(SSL *, EVP_PKEY **);
+
+
+/* cert callbacks were added in OpenSSL 1.0.2 */
+void SSL_set_cert_cb(SSL *s, int (*cert_cb)(SSL *ssl, void *arg), void *arg);
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/ssl.py]: `cerb_cb` should be removed here, as should all the other vairable names.

File: src/cryptography/hazmat/bindings/openssl/ssl.py [hunk 5]
```diff
@@ -387,6 +407,13 @@
...
+void SSL_CTX_set_cert_cb(SSL_CTX *c,
+                         int (*cert_cb)(SSL *ssl, void *arg),
+                         void *arg);
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/ssl.py]: Same here.

File: src/cryptography/hazmat/bindings/openssl/engine.py [hunk 6]
```diff
@@ -22,6 +22,7 @@
 typedef ... *ENGINE_GEN_INT_FUNC_PTR;
 typedef ... *ENGINE_CTRL_FUNC_PTR;
 typedef ... *ENGINE_LOAD_KEY_PTR;
+typedef ... *ENGINE_SSL_CLIENT_CERT_PTR;
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/engine.py]: This is unused, do we care? (Actually all changes to the file are noops)

Comment (reaperhulk) [inline src/cryptography/hazmat/bindings/openssl/engine.py]: We'll need it for that next PR so seems like no big deal to include it.

Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/engine.py]: k

On Sat, May 16, 2015 at 5:24 PM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/bindings/openssl/engine.py
> <https://github.com/pyca/cryptography/pull/1965#discussion_r30464912>:
>
> > @@ -22,6 +22,7 @@
> >  typedef ... *ENGINE_GEN_INT_FUNC_PTR;
> >  typedef ... *ENGINE_CTRL_FUNC_PTR;
> >  typedef ... *ENGINE_LOAD_KEY_PTR;
> > +typedef ... *ENGINE_SSL_CLIENT_CERT_PTR;
>
> We'll need it for that next PR so seems like no big deal to include it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1965/files#r30464912>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by alex** [comment]
lgtm, will merge assuming tests pass

**Comment by alex** [comment]
I don't really understand what's happening, but CentOS is not down with this patch.

**Comment by reaperhulk** [comment]
Tried to simplify this even more, still broken. Awesome.

**Comment by alex** [comment]
Can you file a bug on py.test?

**Comment by alex** [comment]
@reaperhulk There's a new pytest release which fixes some internal errors, try rebasing?

**Comment by alex** [comment]
Jenkins, retest this please

**Comment by reaperhulk** [comment]
retest this please

**Pull request merged**